v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerShardedMigrationIT.java

```diff
@@ -0,0 +1,400 @@
+                        jobInfo2,
+                        TABLE,
+                        "Users-backfill-logical-shard3.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Users-backfill-logical-shard3.jsonl"),
+                    uploadDataStreamFile(
+                        jobInfo2,
+                        TABLE,
+                        "Users-backfill-logical-shard4.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Users-backfill-logical-shard4.jsonl"),
+                    uploadDataStreamFile(
+                        jobInfo2,
+                        TABLE,
+                        "Users-cdc-shard2.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Users-cdc-shard2.jsonl")))
+            .build();
+
+    result =
+        pipelineOperator()
+            .waitForCondition(createConfig(jobInfo2, Duration.ofMinutes(8)), conditionCheck);
+    assertThatResult(result).meetsConditions();
+
+    ConditionCheck rowsConditionCheck =
+        SpannerRowsCheck.builder(spannerResourceManager, TABLE)
+            .setMinRows(12)
+            .setMaxRows(12)
+            .build();
+    result =
+        pipelineOperator()
+            .waitForCondition(createConfig(jobInfo1, Duration.ofMinutes(10)), rowsConditionCheck);
+    assertThatResult(result).meetsConditions();
+
+    // Assert specific rows
+    assertUsersTableContents();
+  }
+
+  @Test
+  public void pkReorderedMultiShardMigration() {
+    // Construct a ChainedConditionCheck with 2 stages.
+    // 1. Send initial wave of events
+    // 2. Wait on Spanner to have events
+    ChainedConditionCheck conditionCheck =
+        ChainedConditionCheck.builder(
+                List.of(
+                    uploadDataStreamFile(
+                        jobInfo1,
+                        MOVIE_TABLE,
+                        "Movie-shard1.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Movie-shard1.jsonl"),
+                    uploadDataStreamFile(
+                        jobInfo1,
+                        MOVIE_TABLE,
+                        "Movie-shard2.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Movie-shard2.jsonl")))
+            .build();
+
+    // Wait for conditions
+    PipelineOperator.Result result =
+        pipelineOperator()
+            .waitForCondition(createConfig(jobInfo1, Duration.ofMinutes(8)), conditionCheck);
+
+    // Assert Conditions
+    assertThatResult(result).meetsConditions();
+
+    ConditionCheck rowsConditionCheck =
+        SpannerRowsCheck.builder(spannerResourceManager, MOVIE_TABLE)
+            .setMinRows(6)
+            .setMaxRows(6)
+            .build();
+    result =
+        pipelineOperator()
+            .waitForCondition(createConfig(jobInfo1, Duration.ofMinutes(8)), rowsConditionCheck);
+    assertThatResult(result).meetsConditions();
+
+    // Assert specific rows
+    assertMovieTableContents();
```

Comment:
We should assert the transformation context, that the rows have a new column for the shardId and it contains the expected value of the shard

Reply:
Yes. The all the rows in the table are asserted.